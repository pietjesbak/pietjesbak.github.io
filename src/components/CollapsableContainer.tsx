import classNames from "classnames";
import * as React from "react";
import { Container } from "./Container";

export interface State {
  collapsed: boolean;
}

export enum headerState {
  /**
   * The header is always shown. (default)
   */
  SHOWN,

  /**
   * The header is shown when hovering inside the container.
   */
  SHOWN_ON_HOVER,

  /**
   * The header is shown when hovering inside the container and always shown when collapsed.
   */
  SHOWN_COLLAPSED,
}

export interface Props extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Is the container open or closed by default.
   * Default open.
   */
  defaultCollapsed?: boolean;

  /**
   * An optional title to show when the container is closed.
   */
  title?: string;

  /**
   * How should the header behave?
   * Default shown.
   */
  header?: headerState;

  /**
   * Does the container contain an error? Affects the colors.
   * Note: this prop is just passed through to the child container.
   */
  error?: boolean;

  /**
   * Store the collapsed state in localstorage? Default false
   * A title is required to store the state.
   */
  storeCollapsed?: boolean;
}

export class CollapsableContainer extends React.Component<Props, State> {
  static defaultProps = {
    header: headerState.SHOWN,
  };

  constructor(props: Props) {
    super(props);

    let collapsed: boolean | null = null;
    if (props.storeCollapsed && props.title) {
      const stored = localStorage.getItem(encodeURIComponent(props.title));
      if (stored !== null) {
        collapsed = stored === "true";
      }
    }
    if (collapsed === null) {
      collapsed = props.defaultCollapsed || false;
    }

    this.state = {
      collapsed,
    };
  }

  toggle = () => {
    const newState = !this.state.collapsed;
    this.setState({
      collapsed: newState,
    });

    if (this.props.storeCollapsed && this.props.title) {
      localStorage.setItem(
        encodeURIComponent(this.props.title),
        String(newState)
      );
    }
  };

  render(): JSX.Element {
    return this.state.collapsed
      ? this.renderCollapsed()
      : this.renderExpanded();
  }

  renderCollapsed() {
    const { children, className, title, header, storeCollapsed, ...rest } =
      this.props;
    return (
      <Container className={classNames(className, "collapsable")} {...rest}>
        <h3
          className={classNames("collapsed-header", {
            hover: header === headerState.SHOWN_ON_HOVER,
          })}
          onClick={this.toggle}
        >
          <i className="container-button icon-window-maximize" />
          <span>{title}</span>
        </h3>
      </Container>
    );
  }

  renderExpanded() {
    const { children, className, title, header, storeCollapsed, ...rest } =
      this.props;

    return (
      <Container className={classNames(className, "collapsable")} {...rest}>
        <h3
          className={classNames("collapsed-header", {
            hover:
              header === headerState.SHOWN_ON_HOVER ||
              header === headerState.SHOWN_COLLAPSED,
          })}
          onClick={this.toggle}
        >
          <i className="container-button icon-window-minimize" />
          <span>{title}</span>
        </h3>
        <div className="content">{children}</div>
      </Container>
    );
  }
}
