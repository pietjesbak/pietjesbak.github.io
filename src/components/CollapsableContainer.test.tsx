import * as enzyme from "enzyme";
import * as React from "react";

import { CollapsableContainer, headerState } from "./CollapsableContainer";

function storageMock() {
  let storage = {};

  return {
    setItem(key: string, value: any) {
      storage[key] = value || "";
    },
    getItem(key: string) {
      return key in storage ? storage[key] : null;
    },
    removeItem(key: string) {
      delete storage[key];
    },
    get length() {
      return Object.keys(storage).length;
    },
    key(i: string) {
      let keys = Object.keys(storage);
      return keys[i] || null;
    },
  };
}

(window as any).localStorage = storageMock();

describe("<CollapsableContainer />", () => {
  it("Renders the default collapsable container", () => {
    const container = enzyme.shallow(
      <CollapsableContainer>CONTENT</CollapsableContainer>
    );

    expect(container).toMatchSnapshot();
    expect(container.find(".content")).toHaveLength(1);

    // Toggle
    container.find("h3").simulate("click");
    expect(container).toMatchSnapshot();
    expect(container.find(".content")).toHaveLength(0);
  });

  it("Renders a collapsable container with settings", () => {
    const container = enzyme.shallow(
      <CollapsableContainer defaultCollapsed={true} title="container">
        CONTENT
      </CollapsableContainer>
    );

    expect(container).toMatchSnapshot();
    expect(container.find(".content")).toHaveLength(0);

    // Toggle
    container.find("h3").simulate("click");
    expect(container).toMatchSnapshot();
    expect(container.find(".content")).toHaveLength(1);
  });

  it("Loads the state from localstorage", () => {
    (window as any).localStorage.setItem("container", "true");
    const container = enzyme.shallow(
      <CollapsableContainer storeCollapsed={true} title="container">
        CONTENT
      </CollapsableContainer>
    );

    expect(container.find(".content")).toHaveLength(0);

    // Toggle
    container.find("h3").simulate("click");
    expect(container.find(".content")).toHaveLength(1);
    expect((window as any).localStorage.getItem("container")).toBe("false");
  });

  it("Gets the error style", () => {
    const container = enzyme.mount(
      <CollapsableContainer error={true} title="container">
        CONTENT
      </CollapsableContainer>
    );

    expect(container.find("div").first().prop("className")).toBe(
      "collapsable container error"
    );
    container.unmount();
  });

  it("Gets the correct classes for each headerstate", () => {
    const containersWithHover = [
      <CollapsableContainer
        key="1"
        header={headerState.SHOWN_ON_HOVER}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
      <CollapsableContainer
        key="2"
        header={headerState.SHOWN_ON_HOVER}
        defaultCollapsed={true}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
      <CollapsableContainer
        key="3"
        header={headerState.SHOWN_COLLAPSED}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
    ];

    const containersWithoutHover = [
      <CollapsableContainer key="4" title="container">
        CONTENT
      </CollapsableContainer>,
      <CollapsableContainer
        key="5"
        header={headerState.SHOWN}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
      <CollapsableContainer
        key="6"
        header={headerState.SHOWN}
        defaultCollapsed={true}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
      <CollapsableContainer
        key="7"
        header={headerState.SHOWN_COLLAPSED}
        defaultCollapsed={true}
        title="container"
      >
        CONTENT
      </CollapsableContainer>,
    ];

    containersWithHover.forEach((component) => {
      const container = enzyme.shallow(component);
      expect(container.find("h3").prop("className")).toBe(
        "collapsed-header hover"
      );
    });

    containersWithoutHover.forEach((component) => {
      const container = enzyme.shallow(component);
      expect(container.find("h3").prop("className")).toBe("collapsed-header");
    });
  });
});
