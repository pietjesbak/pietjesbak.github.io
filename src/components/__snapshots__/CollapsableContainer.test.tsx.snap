// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollapsableContainer /> Renders a collapsable container with settings 1`] = `
<Container
  className="collapsable"
  defaultCollapsed={true}
>
  <h3
    className="collapsed-header"
    onClick={[Function]}
  >
    <i
      className="container-button icon-window-maximize"
    />
    <span>
      container
    </span>
  </h3>
</Container>
`;

exports[`<CollapsableContainer /> Renders a collapsable container with settings 2`] = `
<Container
  className="collapsable"
  defaultCollapsed={true}
>
  <h3
    className="collapsed-header"
    onClick={[Function]}
  >
    <i
      className="container-button icon-window-minimize"
    />
    <span>
      container
    </span>
  </h3>
  <div
    className="content"
  >
    CONTENT
  </div>
</Container>
`;

exports[`<CollapsableContainer /> Renders the default collapsable container 1`] = `
<Container
  className="collapsable"
>
  <h3
    className="collapsed-header"
    onClick={[Function]}
  >
    <i
      className="container-button icon-window-minimize"
    />
    <span />
  </h3>
  <div
    className="content"
  >
    CONTENT
  </div>
</Container>
`;

exports[`<CollapsableContainer /> Renders the default collapsable container 2`] = `
<Container
  className="collapsable"
>
  <h3
    className="collapsed-header"
    onClick={[Function]}
  >
    <i
      className="container-button icon-window-maximize"
    />
    <span />
  </h3>
</Container>
`;
